streaming/src/java/org/apache/hive/streaming/AbstractRecordWriter.java

```diff
@@ -514,15 +575,27 @@ protected RecordUpdater getRecordUpdater(List<String> partitionValues, int bucke
           PartitionInfo partitionInfo = conn.createPartitionIfNotExists(partitionValues);
           // collect the newly added partitions. connection.commitTransaction() will report the dynamically added
           // partitions to TxnHandler
-          if (!partitionInfo.isExists()) {
+          switch(partitionInfo.getExists()) {
+          case NOT_EXISTS:
             addedPartitions.add(partitionInfo.getName());
             if (LOG.isDebugEnabled()) {
               LOG.debug("Created partition {} for table {}", partitionInfo.getName(), fullyQualifiedTableName);
             }
-          } else {
+            break;
+          case EXISTS:
             if (LOG.isDebugEnabled()) {
               LOG.debug("Partition {} already exists for table {}", partitionInfo.getName(), fullyQualifiedTableName);
             }
+            break;
+          case UNKNOWN:
+            undeterminedPartitions.add(partitionInfo.getName());
```

Comment:
looks like the undetermined partitions are sent by workers which does not talk to metastore to know whether a partition exists or not. I am not sure if we really need UNKNOWN state. Workers can send set of partitions to which it added data regardless of whether the partition exists or not. Driver can gather all the partitions from workers, consolidate it and commit with partitions. Does that work?